File: flask/Ogerpon.py
from bs4 import BeautifulSoup
import requests
import json

def getPokemonInfo(criteria, pokemon):
    request = requests.get("https://pokeapi.co/api/v2/" + criteria + "/" + str(pokemon) + "/")
    json_string = request.text
    data = json.loads(json_string)
    return data

def calcBaseStat(*args):
    base_stat = 0
    for arg in args:
        base_stat +=arg
    return base_stat

def getPokemonName(pokemon):
    pokemon = getPokemonInfo("pokemon", pokemon)
    return pokemon["name"]

def getPokemonType(pokemon):
    types = []
    pokemon = getPokemonInfo("pokemon", pokemon)
    for type in pokemon["types"]:
        types.append(type["type"]["name"])
    # return pokemon["types"][0]["type"]["name"]
    return types

def pokemonStats(pokemon):
    pokemon_stats = getPokemonInfo("pokemon", pokemon)
    return {
        'hp': pokemon_stats["stats"][0]['base_stat'],
        'attack': pokemon_stats["stats"][1]['base_stat'],
        'defense': pokemon_stats["stats"][2]['base_stat'],
        'sp. attack': pokemon_stats["stats"][3]['base_stat'],
        'sp. defense': pokemon_stats["stats"][4]['base_stat'],
        'speed': pokemon_stats["stats"][5]['base_stat'],
        'base_stat': calcBaseStat(pokemon_stats["stats"][0]['base_stat'], pokemon_stats["stats"][1]['base_stat'], pokemon_stats["stats"][2]['base_stat'], pokemon_stats["stats"][3]['base_stat'], pokemon_stats["stats"][4]['base_stat'], pokemon_stats["stats"][5]['base_stat'])
    }


def getPokemonAbilities(pokemon):
    slot=0
    abilities = {
        'ability': ['none', 'none'],
        'hidden_ability': 'none'
    }
    pokemon = getPokemonInfo("pokemon", pokemon)
    for ability in pokemon["abilities"]:
        if ability["is_hidden"] is True:
            abilities["hidden_ability"] = ability["ability"]["name"]
        else:
            abilities["ability"][slot] = ability["ability"]["name"]
            slot+=1
    return abilities

def getMove(move, config):
    request = requests.get("https://pokeapi.co/api/v2/move/" + str(move) + "/")
    data = request.json()
    if config == 'name':
        move_info = {
            "name": data["name"],
        }
    elif config == 'basic':
        move_info = {
            "name": data["name"],
            "type": data["type"]["name"],
            "damage_class": data["damage_class"]["name"],
            "power": data["power"],
        }
    elif config == 'all':
        move_info = {
            "name": data["name"],
            "type": data["type"]["name"],
            "damage_class": data["damage_class"]["name"],
            "power": data["power"],
            "accuracy": data["accuracy"],
            "pp": data["pp"],
            "priority": data["priority"],
            "effect": data["effect_entries"][0]["short_effect"] if data["effect_entries"] else None
        }
    return move_info
    
    
# Need to add [:5], otherwise all moves are loaded causing a heavy API call.
def getPokemonMoves(pokemon, config):
    move_list = []
    pokemon_info = getPokemonInfo("pokemon", pokemon)
    for move in pokemon_info["moves"][:5]:
        move_list.append(getMove(move["move"]["name"], config))

    return move_list
        
# Gets the current pokemons pre-evolution or post evolution, if either exist, otherwise displays 'None'
# Parameter(Integer or String)
def getEvolution(pokemon):
    evolvesFrom = requests.get("https://pokeapi.co/api/v2/pokemon-species/" + str(pokemon) + "/")
    evo_data = evolvesFrom.json()
    # return data["chain"]["evolves_to"][0]["species"]["name"]
    evolution_chain = {
        'pre_evo': 'None',
    }
    print(evo_data)
    if evo_data["evolves_from_species"] is None:
        return "None"
    else:
        evolution_chain["pre_evo"] = evo_data["evolves_from_species"]["name"]

    # if post_evo_data["chain"]
    return evolution_chain

# This function requires the initial evoluton chains id (e.g. Venasaur chain -> 1, Charizard -> 2, Blastoise -> 3)
# Parameter(Integer)
def getPokemonEvolutionChain(chainId):
    response = requests.get("https://pokeapi.co/api/v2/evolution-chain/" + str(chainId) + "/")
    evo_data = response.json()
    evo_chain = {
        0: "None",
        1: "None",
        2: "None"
    }
    # Stores the initial pokemon in the evolution chain.
    evo_chain[0] = evo_data["chain"]["species"]
    # Stores the initial pokemon in the evolution chain. If a pokemon does not,
    # continue the chain it returns the object with one pokemon.
    if evo_data["chain"]["evolves_to"]:
        print(evo_data["chain"]["evolves_to"][0]["species"]) 
        evo_chain[1] = evo_data["chain"]["evolves_to"][0]["species"]

    # Stores the initial pokemon in the evolution chain. If a pokemon does not,
    # continue the chain it returns the object with two pokemon.
    if evo_data["chain"]["evolves_to"][0]["evolves_to"] :
        print(evo_data["chain"]["evolves_to"][0]["evolves_to"][0]["species"])
        evo_chain[2] = evo_data["chain"]["evolves_to"][0]["evolves_to"][0]["species"]

    return evo_chain


# getPokemonMoves func works, but takes takes a while before fetching limited number of moves.
def getPokemon(pokemon):
    name = getPokemonName(pokemon)
    types = getPokemonType(pokemon)
    ability = getPokemonAbilities(pokemon)
    stats = pokemonStats(pokemon)
    # moves = getPokemonMoves(pokemon, 'name')
    evo = getPokemonEvolutionChain(pokemon)
    print("Name: " + name)
    print("Type: " + str(types))
    print("Abilities: " + str(ability))
    print("Stats: " + str(stats))
    # print("Moves: " + str(moves))
    print("Evolutions: " + str(evo))
    return {
        "name": name,
        "type": types,
        "ability": ability,
        "stats": stats,
        "evo": evo
    }

getPokemon(134)

File: node/Ogerpon.js
async function getPokemonInfo(criteria, pokemon) {
  const response = await fetch(`https://pokeapi.co/api/v2/${criteria}/${pokemon}/`);
  const data = await response.json();
  return data;
}

function calcBaseStat(...args) {
  return args.reduce((total, stat) => total + stat, 0);
}

async function getPokemonName(pokemon) {
  const data = await getPokemonInfo("pokemon", pokemon);
  return data.name;
}

async function getPokemonType(pokemon) {
  const data = await getPokemonInfo("pokemon", pokemon);
  return data.types.map(t => t.type.name);
}

async function pokemonStats(pokemon) {
  const statsData = await getPokemonInfo("pokemon", pokemon);
  const stats = statsData.stats;
  return {
    hp: stats[0].base_stat,
    attack: stats[1].base_stat,
    defense: stats[2].base_stat,
    'sp. attack': stats[3].base_stat,
    'sp. defense': stats[4].base_stat,
    speed: stats[5].base_stat,
    base_stat: calcBaseStat(
      stats[0].base_stat,
      stats[1].base_stat,
      stats[2].base_stat,
      stats[3].base_stat,
      stats[4].base_stat,
      stats[5].base_stat
    )
  };
}

async function getPokemonAbilities(pokemon) {
  const data = await getPokemonInfo("pokemon", pokemon);
  const abilities = {
    ability: ['none', 'none'],
    hidden_ability: 'none'
  };
  let slot = 0;
  data.abilities.forEach(ability => {
    if (ability.is_hidden) {
      abilities.hidden_ability = ability.ability.name;
    } else {
      abilities.ability[slot] = ability.ability.name;
      slot++;
    }
  });
  return abilities;
}

async function getMove(move, config) {
  const res = await fetch(`https://pokeapi.co/api/v2/move/${move}/`);
  const data = await res.json();
  if (config === 'name') {
    return { name: data.name };
  } else if (config === 'basic') {
    return {
      name: data.name,
      type: data.type.name,
      damage_class: data.damage_class.name,
      power: data.power
    };
  } else if (config === 'all') {
    return {
      name: data.name,
      type: data.type.name,
      damage_class: data.damage_class.name,
      power: data.power,
      accuracy: data.accuracy,
      pp: data.pp,
      priority: data.priority,
      effect: data.effect_entries.length > 0 ? data.effect_entries[0].short_effect : null
    };
  }
}

async function getPokemonMoves(pokemon, config) {
  const data = await getPokemonInfo("pokemon", pokemon);
  const moves = await Promise.all(
    data.moves.slice(0, 5).map(m => getMove(m.move.name, config))
  );
  return moves;
}

async function getEvolution(pokemon) {
  const res = await fetch(`https://pokeapi.co/api/v2/pokemon-species/${pokemon}/`);
  const data = await res.json();
  const evolution_chain = {
    pre_evo: 'None'
  };
  if (data.evolves_from_species) {
    evolution_chain.pre_evo = data.evolves_from_species.name;
  } else {
    return 'None';
  }
  return evolution_chain;
}

async function getPokemonEvolutionChain(chainId) {
  const res = await fetch(`https://pokeapi.co/api/v2/evolution-chain/${chainId}/`);
  const data = await res.json();
  const evo_chain = {
    0: "None",
    1: "None",
    2: "None"
  };
  evo_chain[0] = data.chain.species;
  if (data.chain.evolves_to.length > 0) {
    evo_chain[1] = data.chain.evolves_to[0].species;
    if (data.chain.evolves_to[0].evolves_to.length > 0) {
      evo_chain[2] = data.chain.evolves_to[0].evolves_to[0].species;
    }
  }
  return evo_chain;
}

export default async function getPokemon(pokemon) {
  const name = await getPokemonName(pokemon);
  const types = await getPokemonType(pokemon);
  const ability = await getPokemonAbilities(pokemon);
  const stats = await pokemonStats(pokemon);
  const evo = await getPokemonEvolutionChain(pokemon); // should be evolution-chain ID, not pokemon ID

  console.log("Name: " + name);
  console.log("Type: " + JSON.stringify(types));
  console.log("Abilities: " + JSON.stringify(ability));
  console.log("Stats: " + JSON.stringify(stats));
  console.log("Evolutions: " + JSON.stringify(evo));

  return {
    name,
    type: types,
    ability,
    stats,
    evo
  };
}

// Example usage:
getPokemon(134);
